File: lib/swimImportExport.js
const EXPORT_VERSION = 3;

const CRC32_TABLE = (() => {
  const table = [];
  for (let n = 0; n < 256; n++) {
    let c = n;
    for (let k = 0; k < 8; k++) {
      c = c & 1 ? 0xedb88320 ^ (c >>> 1) : c >>> 1;
    }
    table[n] = c;
  }
  return table;
})();

const crc32 = (str) => {
  let crc = 0 ^ -1;
  for (let i = 0; i < str.length; i++) {
    crc = (crc >>> 8) ^ CRC32_TABLE[(crc ^ str.charCodeAt(i)) & 0xff];
  }
  return ((crc ^ -1) >>> 0).toString(16).padStart(8, '0');
};

const bytesToBinaryString = (bytes) => {
  let binary = '';
  for (let i = 0; i < bytes.length; i++) {
    binary += String.fromCharCode(bytes[i]);
  }
  return binary;
};

const binaryStringToBytes = (binary) => {
  const bytes = new Uint8Array(binary.length);
  for (let i = 0; i < binary.length; i++) {
    bytes[i] = binary.charCodeAt(i);
  }
  return bytes;
};

const encodeBase64FromBytes = (bytes) => {
  if (typeof Buffer !== 'undefined') {
    return Buffer.from(bytes).toString('base64');
  }
  return btoa(bytesToBinaryString(bytes));
};

const decodeBase64ToBytes = (encoded) => {
  if (typeof Buffer !== 'undefined') {
    return Uint8Array.from(Buffer.from(encoded, 'base64'));
  }
  return binaryStringToBytes(atob(encoded));
};

const encodeUtf8Json = (jsonStr) => {
  if (typeof Buffer !== 'undefined') {
    return encodeBase64FromBytes(Buffer.from(jsonStr, 'utf8'));
  }
  return btoa(unescape(encodeURIComponent(jsonStr)));
};

const decodeUtf8Json = (encoded) => {
  if (typeof Buffer !== 'undefined') {
    return Buffer.from(encoded, 'base64').toString('utf8');
  }
  return decodeURIComponent(escape(atob(encoded)));
};

const isGzipBytes = (bytes) => bytes.length >= 2 && bytes[0] === 0x1f && bytes[1] === 0x8b;

const tryGzip = async (text) => {
  if (typeof CompressionStream === 'undefined') return null;
  const input = new Blob([new TextEncoder().encode(text)]).stream();
  const compressedStream = input.pipeThrough(new CompressionStream('gzip'));
  const buffer = await new Response(compressedStream).arrayBuffer();
  return new Uint8Array(buffer);
};

const tryGunzip = async (bytes) => {
  if (!isGzipBytes(bytes) || typeof DecompressionStream === 'undefined') return null;
  const decompressedStream = new Blob([bytes]).stream().pipeThrough(new DecompressionStream('gzip'));
  return new Response(decompressedStream).text();
};

const encodePayloadJson = async (jsonStr) => {
  const gzipped = await tryGzip(jsonStr);
  if (gzipped) return encodeBase64FromBytes(gzipped);
  return encodeUtf8Json(jsonStr);
};

const decodePayloadJson = async (encoded) => {
  const bytes = decodeBase64ToBytes(encoded);
  const gunzipped = await tryGunzip(bytes);
  if (gunzipped != null) return gunzipped;
  return decodeUtf8Json(encoded);
};

const compressSession = (session) => ({
  i: session.id,
  dt: session.date,
  ca: session.createdAt,
  m: {
    ds: session.metrics?.durationSec,
    dm: session.metrics?.distanceM,
    ak: session.metrics?.activeKcal,
    tk: session.metrics?.totalKcal,
    ps: session.metrics?.paceSecPer100m,
    hr: session.metrics?.avgHeartRate,
    lp: session.metrics?.laps,
    pl: session.metrics?.poolLengthM,
    gm: session.metrics?.goalM,
    loc: session.metrics?.location,
    tr: session.metrics?.timeRange,
    st: session.metrics?.strokes,
  },
});

const decompressSession = (raw) => ({
  id: raw.i,
  date: raw.dt,
  createdAt: raw.ca,
  metrics: {
    durationSec: raw.m?.ds ?? null,
    distanceM: raw.m?.dm ?? null,
    activeKcal: raw.m?.ak ?? null,
    totalKcal: raw.m?.tk ?? null,
    paceSecPer100m: raw.m?.ps ?? null,
    avgHeartRate: raw.m?.hr ?? null,
    laps: raw.m?.lp ?? null,
    poolLengthM: raw.m?.pl ?? 25,
    goalM: raw.m?.gm ?? null,
    location: raw.m?.loc ?? '',
    timeRange: raw.m?.tr ?? '',
    strokes: raw.m?.st ?? {},
  },
});

export const generateSwimExportString = async (swimData) => {
  const payload = {
    v: EXPORT_VERSION,
    p: swimData.profile,
    s: (swimData.sessions || []).map(compressSession),
  };
  const jsonStr = JSON.stringify(payload);
  const encoded = await encodePayloadJson(jsonStr);
  const checksum = crc32(encoded);
  return `${encoded}:${checksum}`;
};

export const parseSwimImportString = async (str) => {
  if (!str || typeof str !== 'string') {
    throw new Error('Invalid input');
  }

  const parts = str.trim().split(':');
  if (parts.length !== 2) {
    throw new Error('Invalid format');
  }

  const [encoded, checksum] = parts;
  const expectedChecksum = crc32(encoded);
  if (expectedChecksum !== checksum.toLowerCase()) {
    throw new Error('Checksum validation failed');
  }

  const decoded = await decodePayloadJson(encoded);
  const compressed = JSON.parse(decoded);

  if (compressed.v !== EXPORT_VERSION) {
    throw new Error('Unsupported export version');
  }

  return {
    profile: compressed.p || { sex: 'male', age: 30 },
    sessions: (compressed.s || []).map(decompressSession),
  };
};

/** @internal */
export const __testing = { EXPORT_VERSION, compressSession, decompressSession, crc32 };

File: ios/AapSC/Lib/SwimImportExport.swift
import Foundation
import Compression

enum SwimImportExport {
    static let exportVersion = 10

    private static let crc32Table: [UInt32] = {
        (0..<256).map { n -> UInt32 in
            var c = UInt32(n)
            for _ in 0..<8 {
                c = (c & 1) != 0 ? (0xEDB88320 ^ (c >> 1)) : (c >> 1)
            }
            return c
        }
    }()

    static func generateExportString(from data: SwimData) async throws -> String {
        let payload: [String: Any] = [
            "v": exportVersion,
            "p": [
                "sex": data.profile.sex,
                "age": data.profile.age,
            ],
            "s": data.sessions.map(compressSession),
            "aa": data.profile.activeAmbient as Any,
        ]
        let jsonData = try JSONSerialization.data(withJSONObject: payload)
        guard let jsonStr = String(data: jsonData, encoding: .utf8) else {
            throw SwimImportExportError.invalidInput
        }
        let encoded = try await encodePayloadJson(jsonStr)
        let checksum = crc32(encoded)
        return "\(encoded):\(checksum)"
    }

    static func parseImportString(_ str: String) async throws -> SwimData {
        let trimmed = str.trimmingCharacters(in: .whitespacesAndNewlines)
        let parts = trimmed.split(separator: ":", maxSplits: 1).map(String.init)
        guard parts.count == 2 else { throw SwimImportExportError.invalidFormat }

        let encoded = parts[0]
        let checksum = parts[1].lowercased()
        guard crc32(encoded) == checksum else { throw SwimImportExportError.checksumFailed }

        let decoded = try await decodePayloadJson(encoded)
        guard let compressed = try JSONSerialization.jsonObject(with: Data(decoded.utf8)) as? [String: Any] else {
            throw SwimImportExportError.invalidFormat
        }

        let version = compressed["v"] as? Int ?? 0
        guard [3, 4, 5, 6, 7, 8, 9, 10].contains(version) else {
            throw SwimImportExportError.unsupportedVersion
        }

        let profile = SwimProfile(
            name: "",
            sex: (compressed["p"] as? [String: Any])?["sex"] as? String ?? "male",
            age: (compressed["p"] as? [String: Any])?["age"] as? Int ?? 30,
            mascotId: nil,
            mascotSwitchMonthKey: nil,
            aiApiKey: "",
            activeAmbient: version >= 6 ? (compressed["aa"] as? String) : nil
        )

        let sessions = (compressed["s"] as? [[String: Any]] ?? []).map(decompressSession)

        return SwimData(
            profile: profile,
            sessions: sessions,
            monthlyChallengeRerolls: [:]
        )
    }

    enum SwimImportExportError: LocalizedError {
        case invalidInput
        case invalidFormat
        case checksumFailed
        case unsupportedVersion

        var errorDescription: String? {
            switch self {
            case .invalidInput: return "Invalid input"
            case .invalidFormat: return "Invalid format"
            case .checksumFailed: return "Checksum validation failed"
            case .unsupportedVersion: return "Unsupported export version"
            }
        }
    }

    private static func crc32(_ str: String) -> String {
        var crc: UInt32 = 0xFFFF_FFFF
        for byte in str.utf8 {
            crc = (crc >> 8) ^ crc32Table[Int((crc ^ UInt32(byte)) & 0xFF)]
        }
        return String(format: "%08x", (crc ^ 0xFFFF_FFFF))
    }

    private static func encodePayloadJson(_ jsonStr: String) async throws -> String {
        Data(jsonStr.utf8).base64EncodedString()
    }

    private static func decodePayloadJson(_ encoded: String) async throws -> String {
        guard let bytes = Data(base64Encoded: encoded) else {
            throw SwimImportExportError.invalidFormat
        }
        if bytes.count >= 2, bytes[0] == 0x1f, bytes[1] == 0x8b,
           let gunzipped = gzipDecompress(bytes),
           let text = String(data: gunzipped, encoding: .utf8) {
            return text
        }
        guard let text = String(data: bytes, encoding: .utf8) else {
            throw SwimImportExportError.invalidFormat
        }
        return text
    }

    private static func gzipDecompress(_ data: Data) -> Data? {
        decompress(data, algorithm: COMPRESSION_ZLIB)
    }

    private static func compress(_ data: Data, algorithm: compression_algorithm) -> Data? {
        data.withUnsafeBytes { buffer in
            let dstSize = max(64, data.count + 64)
            var dst = Data(count: dstSize)
            let written = dst.withUnsafeMutableBytes { dstBuffer -> Int in
                guard let srcPtr = buffer.baseAddress?.assumingMemoryBound(to: UInt8.self),
                      let dstPtr = dstBuffer.baseAddress?.assumingMemoryBound(to: UInt8.self) else {
                    return 0
                }
                return compression_encode_buffer(
                    dstPtr, dstBuffer.count,
                    srcPtr, buffer.count,
                    nil, algorithm
                )
            }
            guard written > 0 else { return nil }
            dst.count = written
            return dst
        }
    }

    private static func decompress(_ data: Data, algorithm: compression_algorithm) -> Data? {
        data.withUnsafeBytes { buffer in
            var dst = Data(count: max(data.count * 4, 1024))
            var written = 0
            while written == 0 && dst.count <= data.count * 32 {
                written = dst.withUnsafeMutableBytes { dstBuffer -> Int in
                    guard let srcPtr = buffer.baseAddress?.assumingMemoryBound(to: UInt8.self),
                          let dstPtr = dstBuffer.baseAddress?.assumingMemoryBound(to: UInt8.self) else {
                        return 0
                    }
                    return compression_decode_buffer(
                        dstPtr, dstBuffer.count,
                        srcPtr, buffer.count,
                        nil, algorithm
                    )
                }
                if written == 0 { dst.count *= 2 }
            }
            guard written > 0 else { return nil }
            dst.count = written
            return dst
        }
    }

    static func compressSessionForTesting(_ session: SwimSession) -> [String: Any] {
        compressSession(session)
    }

    static func decompressSessionForTesting(_ raw: [String: Any]) -> SwimSession {
        decompressSession(raw)
    }

    private static func compressSession(_ session: SwimSession) -> [String: Any] {
        var result: [String: Any] = [
            "i": session.id,
            "dt": session.date,
            "m": compressMetrics(session.metrics),
        ]
        if let createdAt = session.createdAt { result["ca"] = createdAt }
        if session.excludeFromStats { result["ex"] = true }
        if let uuid = session.healthKitWorkoutUUID { result["hk"] = uuid }
        return result
    }

    private static func decompressSession(_ raw: [String: Any]) -> SwimSession {
        SwimSession(
            id: raw["i"] as? String ?? UUID().uuidString,
            createdAt: raw["ca"] as? String,
            date: raw["dt"] as? String ?? "",
            metrics: decompressMetrics(raw["m"] as? [String: Any] ?? [:]),
            excludeFromStats: raw["ex"] as? Bool ?? false,
            healthKitWorkoutUUID: raw["hk"] as? String
        )
    }

    private static func compressMetrics(_ metrics: SwimMetrics) -> [String: Any] {
        var result: [String: Any] = [
            "pl": metrics.poolLengthM,
            "loc": metrics.location,
            "tr": metrics.timeRange,
        ]
        if let v = metrics.durationSec { result["ds"] = v }
        if let v = metrics.distanceM { result["dm"] = v }
        if let v = metrics.activeKcal { result["ak"] = v }
        if let v = metrics.totalKcal { result["tk"] = v }
        if let v = metrics.paceSecPer100m { result["ps"] = v }
        if let v = metrics.avgHeartRate { result["hr"] = v }
        if let v = metrics.laps { result["lp"] = v }
        if let v = metrics.goalM { result["gm"] = v }
        return result
    }

    private static func decompressMetrics(_ raw: [String: Any]) -> SwimMetrics {
        SwimMetrics(
            durationSec: raw["ds"] as? Int,
            distanceM: raw["dm"] as? Int,
            activeKcal: raw["ak"] as? Int,
            totalKcal: raw["tk"] as? Int,
            paceSecPer100m: raw["ps"] as? Int,
            avgHeartRate: raw["hr"] as? Int,
            laps: raw["lp"] as? Int,
            poolLengthM: raw["pl"] as? Int ?? 25,
            goalM: raw["gm"] as? Int,
            location: raw["loc"] as? String ?? "",
            timeRange: raw["tr"] as? String ?? "",
            strokes: .empty
        )
    }
}
